src/aeso_ast_infer_types.erl

```diff
@@ -1049,6 +1131,73 @@ check_typedef(Env, {variant_t, Cons}) ->
     {variant_t, [ {constr_t, Ann, Con, [ check_type(Env, Arg) || Arg <- Args ]}
                 || {constr_t, Ann, Con, Args} <- Cons ]}.
 
+infer_type_vars_variance(TypeParams, Cons) ->
+    % args from all type constructors
+    FlatArgs = lists:flatten([Args || {constr_t, _, _, Args} <- Cons]),
+
+    Vs = lists:flatten([infer_type_vars_variance(Arg) || Arg <- FlatArgs]),
+    lists:map(fun({tvar, _, TVar}) ->
+                      S = sets:from_list([Variance || {TV, Variance} <- Vs, TV == TVar]),
+                      IsInvariant     = sets:is_element(invariant, S),
+                      IsCovariant     = sets:is_element(covariant, S),
+                      IsContravariant = sets:is_element(contravariant, S),
+                      IsBivariant     = sets:is_element(bivariant, S),
+                      case {IsInvariant, IsCovariant, IsContravariant, IsBivariant} of
```

Comment:
I think you can just have a set for co and contra and then 
```
case {Co, Contra} of
  {t, f} -> co
  {f, t} -> contra
  {t, t} -> inv
  {f, f} -> bi
```

Reply:
Done here bdd3f0dde2f4423722808cdd5478a7a54d62f85f